pkg/proxy/oxyproxy.go

```diff
@@ -0,0 +1,607 @@
+// Copyright 2016-2017 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package proxy
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"strconv"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/cilium/cilium/common/addressing"
+	"github.com/cilium/cilium/pkg/endpointmanager"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/nodeaddress"
+	"github.com/cilium/cilium/pkg/policy"
+	"github.com/cilium/cilium/pkg/proxy/accesslog"
+
+	"github.com/braintree/manners"
+	log "github.com/sirupsen/logrus"
+	"github.com/vulcand/oxy/forward"
+	"github.com/vulcand/route"
+)
+
+// OxyRedirect implements the Redirect interface for a l7 proxy
+type OxyRedirect struct {
+	mutex lock.RWMutex
+
+	id       string
+	fromPort uint16
+	toPort   uint16
+	epID     uint64
+	rules    []string
+	source   ProxySource
+	server   *manners.GracefulServer
+	router   route.Router
+	l4       policy.L4Filter // stale copy, ignore rules
+	nodeInfo accesslog.NodeAddressInfo
+}
+
+// ToPort returns the redirect port of an OxyRedirect
+func (r *OxyRedirect) ToPort() uint16 {
+	return r.toPort
+}
+
+func (r *OxyRedirect) updateRules(rules []string) {
+	for _, v := range r.rules {
+		r.router.RemoveRoute(v)
+	}
+
+	r.rules = make([]string, len(rules))
+	copy(r.rules, rules)
+
+	for _, v := range r.rules {
+		r.router.AddRoute(v, v)
+	}
+}
+
+func (r *OxyRedirect) localEndpointInfo(info *accesslog.EndpointInfo) {
+	r.source.RLock()
+	info.ID = r.epID
```

Comment:
I asked about the mutex because this doesn't seem to be protected

Reply:
epID remains unchanged, so it does not need protection. I added a separate patch simplifying OxyRedirect and making mutex semantics more explicit. I have already squashed in the more trivial updates, sorry for that...